src/mds/MDCache.cc

```diff
@@ -3555,6 +3555,82 @@ void MDCache::remove_inode_recursive(CInode *in)
   remove_inode(in);
 }
 
+/**
+ * For all inodes, dirs, dentries below an inode, compose expiry
+ * messages.  This is used when giving up all replicas of entities
+ * for an MDS peer in the 'stopping' state, such that the peer can
+ * empty its cache and finish shutting down.
+ *
+ * We have to make sure we're only expiring un-referenced items to
+ * avoid interfering with ongoing stray-movement (we can't distinguish
+ * between the "moving my strays" and "waiting for my cache to empty"
+ * phases within 'stopping')
+ *
+ * @return false if we completed cleanly, true if caller should stop
+ *         expiring because we hit something with refs.
+ */
+bool MDCache::expire_recursive(
+  CInode *in,
+  map<mds_rank_t, MCacheExpire*>& expiremap,
+  CDir *subtree)
```

Comment:
Is there a reason we're passing in and subtree separately here instead of extracting one from the other? It doesn't look to me like things will work properly if they don't match, so we should just be passing one and getting the other from it.